doc/HttpTransportAdapter.md

```diff
@@ -0,0 +1,192 @@
+
+```cpp
+/*
+* Using a session pattern for the HTTP transport adapter.
+*/
+
+// Create a session class first.
+class CustomHttpTransportAdapterSession {
+  private:
+    // Keep any state like network/socket handlers here.
+    SomeHttpClientLibraryHandler handler;
+    SomeOtherRequiredState state;
+  
+  public:
+    // Expose any functionality required. For example, this could
+    // be the main method to produce the HTTP raw response.
+    std::unique_ptr<Azure::Core::Http::RawResponse> Perform(
+      Azure::Core::Context const& context,
+      Azure::Core::Http::Request& request
+    );
+}
+```
+
+Then, within the Send() method implementation.
+
+```cpp
+// Create a new Session object when calling Send().
+    std::unique_ptr<Azure::Core::Http::RawResponse> Send(
+        Azure::Core::Context const& context,
+        Azure::Core::Http::Request& request) {
+
+          // This will ensure that multiple clients can
+          // call Send() at the same time.
+          CustomHttpTransportAdapterSession session;
+
+          // Use any functionally from your created session.
+          // For this example, we could just return like this:
+          return session.Perform();
+        }
+
+```
+
+> If instead of having a session created for every call to Send(), we would place the state directly into the HTTP transport adapter class, any new call to Send() would corrupt the current state when sharing the same HTTP transport adapter from multiple clients.
+
+3. Use the adapter.
+
+At this point you have all you need to use your custom HTTP transport adapter. Refer to [Using the HTTP transport adapter](#using-the-http-transport-adapter) for how to set it upon creating the SDK client.
+
+4. Set as default (optional).
+
+You can optionally set your custom HTTP transport adapter as the default case. This can be helpful if you want to always use this HTTP transport adapter and you would like to avoid setting up the SDK client options every time you create a new SDK client.
+
+The first step is to set the required CMake compile option that would configure the build for using a custom HTTP transport adapter. Refer to the [CMake options](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#cmake-build-options) to find out the required option fot this.
+
+The second step is to implement the method showed below in the global unnamed namespace.
+
+```cpp
+/*
+* Setting an HTTP transport adapter as default.
+*/
+
+// From one of the cpp files in your application:
+std::shared_ptr<Azure::Core::Http::HttpTransport> ::AzureSdkGetCustomHttpTransport() {
+  // Create and return a shared ptr for your custom HTTP transport adapter.
+  return std::make_shared<CustomHttpTransportAdapterSession>();
```

Comment:
If AzureSdkGetCustomHttpTransport is called multiple times, it keeps returning the pointer to a single CustomHttpTransportAdapter, right?
We don't want to create a new adapter with each call.

Also, CustomHttpTransportAdapterSession should be called CustomHttpTransportAdapter, right?

Reply:
We are currently creating a new adapter on every call.
The same happen with curl and winHttp.  When a new SDK client is created with default options, a new transport adapter is created.
The benefit of this is that the transport adapter is destroyed as soon as no SDK client is using it. Otherwise, we would be leaving in memory a global adapter to use for default calls.

We can review this. I will create a new issue to open the discussion on it. It is a good point